spec/models/zip_code_spec.rb

```diff
@@ -3,5 +3,75 @@
 require 'rails_helper'
 
 RSpec.describe ZipCode, type: :model do
-  it { should validate_presence_of(:d_codigo) }
+  context 'validations tests' do
+    it { should validate_presence_of(:d_codigo) }
+  end
+
+  context 'scopes tests' do
+    let(:params) do
+      { d_codigo: '01000',
+        d_asenta: 'San Ángel',
+        d_tipo_asenta: 'Colonia',
+        d_mnpio: 'Álvaro Obregón',
+        d_estado: 'Ciudad de México',
+        d_ciudad: 'Ciudad de México',
+        d_cp: '01001',
+        c_estado: '09',
+        c_oficina: '01001',
+        c_cp: '.',
+        c_tipo_asenta: '09',
+        c_mnpio: '010',
+        id_asenta_cpcons: '0001',
+        d_zona: 'Urbano',
+        c_cve_ciudad: '01' }
+    end
+
+    before(:each) do
+      ZipCode.create(params)
+    end
+
+    it 'should return a zip_code' do
+      expect(ZipCode.count).to eq(1)
+    end
+
+    it 'search for a wrong zip_code gives empty results' do
+      cp = '64000'
+      expect(ZipCode.find_by_zip_code(cp)).to eq([])
+    end
+
+    it 'search for a valid zip_code gives valid results' do
+      cp = '01000'
+      expect(ZipCode.find_by_zip_code(cp)).to eq([ZipCode.last])
+    end
+
+    it 'search for a wrong state gives empty results' do
+      state = 'Nuevo Leon'
+      expect(ZipCode.find_by_state(state)).to eq([])
```

Comment:
This tests has the mystery guess code smell, where you don't actually know where the input data comes from, or at least is not explicit. This test also passes if no record is being created, so watch for that.

Reply:
I don't understand this.

Reply:
All right, this test has two issues:

1. If you don't have any records on the test database, this will always pass, which is expected, but not right, as you really want to test that along with some other records, if you search by `Nuevo Leon` you still won't get a result

In other words, something like:

```ruby
it 'search for a wrong state gives empty results' do
  FactoryBot.create(:zip_code, d_estado: 'Tamaulipas')
  FactoryBot.create(:zip_code, d_estado: 'Coahuila')
  FactoryBot.create(:zip_code, d_estado: 'Sinaloa')

  state_name = 'Nuevo Leon'
  
  expect(ZipCode.find_by_state(state_name)).to be_empty
end
```

Or you can test against records to actually receive, not the opposite as the example above:

```ruby
it 'search for a wrong state gives empty results' do
  nuevo_leon_zip_code_1 = FactoryBot.create(:zip_code, d_estado: 'Nuevo Leon')
  nuevo_leon_zip_code_2 = FactoryBot.create(:zip_code, d_estado: 'Nuevo Leon')
  FactoryBot.create(:zip_code, d_estado: 'Sinaloa')

  state_name = 'Nuevo Leon'
  
  expect(ZipCode.find_by_state(state_name)).to match_array([nuevo_leon_zip_code_1, nuevo_leon_zip_code_2])
end
```

Remember that the `match_array` matcher does not care for the order of the output, which in this case is fine.

2. The second problem this test has is that you are hiding information from the test regarding the creation of the data you are working on, there is no explicit information that you have in this case ZipCodes belonging to another state, which with the examples I provided you can solve this. This is a tricky one, please let me know if I was clear enough.